SUPINT-1785: Describe elastic upgrade (#1604)
nuxeo/doc.nuxeo.com-content: src/nxdoc/nuxeo-server/upgrading-the-nuxeo-platform/upgrade-from-lts-2019-to-cloud-2020.md

## Elasticsearch

> Nuxeo 11.4/Cloud 2021 is using Elasticsearch client library version 7.9.2.

Reviewer: "Nuxeo 11.4/Cloud 2021" seems weird, not sure what terminology we want to use here: Nuxeo 11.4 and above / Nuxeo Cloud release, since 11.4 / ...

Author: ok will use Since Nuxeo 11.4, I guess there will be another review to fix the versionning for cloud/lts.

> once you have adapted your application to follow the behavior changes, 

Reviewer: ```suggestion
Once you have adapted your application to follow the behavior changes, 
```

you need to upgrade your existing Elasticsearch cluster to version 7.9 (7.7 or 7.8 are also supported). 

> ### Behavior changes

Reviewer: ```suggestion
### Behavior Changes
```

#### Update Your Custom Elasticsearch Settings and Mapping

> If you have a custom configuration for Elasticsearch it needs to be adapted to follow the new 7.x rules.

Reviewer: ```suggestion
If you have a custom configuration for Elasticsearch, it needs to be adapted to follow the new 7.x rules.
```

> Here are few breaking changes to take in account when upgrading from Elasticsearch 6:

Reviewer: ```suggestion
Here are a few breaking changes to take into account when upgrading from Elasticsearch 6:
```

- If you use the nGram tokenizer the type has been renamed from `nGram` to `ngram`.

- A phrase prefix search on a `keyword` field now raises an exception, this will happen on the following query:

  ```sql
  SELECT * FROM Document WHERE someKeywordField LIKE 'foo%'
  ```

  The field needs to be of `text` type.

> #### Don't use the TransportClient

Reviewer: ```suggestion
#### Don't Use the TransportClient
```

> The connection to Elasticsearch should be `RestClient` (default since Nuxeo LTS 2017), 

Reviewer: ```suggestion
The connection to Elasticsearch should use the `RestClient` (default since Nuxeo LTS 2017), 
```

the `TransportClient` is deprecated in version 7.x.

> Technically it can work if the index exist, but re-indexing is going to fail.

Reviewer: ```suggestion
Technically, it can work if the index exists, but re-indexing is going to fail.
```

> There is no advantage on using the `TransportClient` over the default `RestClient`. 

Reviewer: ```suggestion
There is no advantage of using the `TransportClient` over the default `RestClient`. 
```

> #### Warnings in Unit Test with embedded Elasticsearch instance

Reviewer: ```suggestion
#### Warnings in Unit Test with Embedded Elasticsearch Instance
```

Your unit tests may have the following WARN messages: 

```
2020-11-16 11:10:43,561 [main] WARN  [DanglingIndicesState] gateway.auto_import_dangling_indices is disabled, dangling indices will not be automatically detected or imported and must be managed manually
2020-11-16 11:10:43,828 [elasticsearch[nuxeoTestNode][clusterApplierService#updateTask][T#1]] WARN  [ClusterApplierService] failed to notify ClusterStateListener
org.apache.lucene.util.SetOnce$AlreadySetException: The object cannot be set twice!
```

You can ignore them by editing your `log4j2` test configuration to add:

> ```

Reviewer: ```suggestion
```xml
```

    <!-- Remove WARN on elastic embedded because dangling indices cannot be detected -->
    <Logger name="org.elasticsearch.gateway.DanglingIndicesState" level="error" />
    <!-- Remove WARN on elastic embedded because node and cluster identifiers change between test suites -->
    <Logger name="org.elasticsearch.cluster.service.ClusterApplierService" level="error" />
```

> #### Adapt your custom Elasticsearch queries

Reviewer: ```suggestion
#### Adapt your Custom Elasticsearch Queries
```

If you use the [Elasticsearch Passthrough]({{page page='elasticsearch-passthrough'}),
or if you use directly the [elasticsearch query builder](https://github.com/nuxeo/nuxeo/blob/v11.4.32/modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/query/NxQueryBuilder.java#L164)

> You have to make sure your query works in Elasticsearch 7.x.

Reviewer: This is the end of the previous sentence, no?

> One of the biggest change is that elasticsearch document type has been removed, 

Reviewer: ```suggestion
One of the biggest changes is that the Elasticsearch document type has been removed, 
```

> for instance, this change the URL pattern to access a Nuxeo document from http://elastic:9200/nuxeo/doc/<DOC_ID> to: http://elastic:9200/nuxeo/<DOC_ID>

Reviewer: ```suggestion
for instance, this changes the URL pattern to access a Nuxeo document from http://elastic:9200/nuxeo/doc/<DOC_ID> to: http://elastic:9200/nuxeo/<DOC_ID>.
```

<i class="fa fa-long-arrow-right" aria-hidden="true"></i>&nbsp;More on JIRA ticket [NXP-21874](https://jira.nuxeo.com/browse/NXP-21874)

### Upgrade Elasticsearch to 7.x

The procedure depends on when your instance has been created:

> #### Elasticsearch indexes created in Elasticsearch 6.x (Nuxeo 10.10/LTS 2019) 

Reviewer: ```suggestion
#### Elasticsearch indexes Created in Elasticsearch 6.x (Nuxeo 10.10/LTS 2019) 
```

Elasticsearch 7.x can read indexes created in version 6.0 or above.

This means that there is no migration to do if your Nuxeo instance has been created with Nuxeo LTS 2019.

Follow the [Elasticsearch upgrade documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/setup-upgrade.html) to upgrade your Elasticsearch Cluster.

> #### Migration of Elastic indices created in Elasticsearch 5.x (Nuxeo 9.10/LTS 2017)

Reviewer: ```suggestion
#### Migration of Elastic Indices Created in Elasticsearch 5.x (Nuxeo 9.10/LTS 2017)
```

 
An Elasticsearch 7.x node will not start in the presence of indices created in a version of Elasticsearch before 6.0.
Indices created in Elasticsearch 5.x or before will need to be reindexed with Elasticsearch 6.x in order to be readable by Elasticsearch 7.x.

Nuxeo uses 3 indexes:

> 1. The repository index named `nuxeo` by default doesn't need this migration because the repository 

Reviewer: ```suggestion
1. The repository index, named `nuxeo` by default, doesn't need this migration because the repository 
```

>  will be re-indexed in the next step, so once this index has been backed up you can delete it.

Reviewer: ```suggestion
 will be re-indexed in the next step, so, once this index has been backed up, you can delete it.
```

> 2. The sequence index named `nuxeo-uidgen` will be re-created at startup, so once this index has been backed up you can delete it.

Reviewer: ```suggestion
2. The sequence index named `nuxeo-uidgen` will be re-created at startup, so, once this index has been backed up, you can delete it.
```

3. The audit index named `nuxeo-audit` needs to be migrated. Follow the [re-index upgrade procedure](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/reindex-upgrade.html).

> Once the Elasticsearch cluster is upgraded, start Nuxeo LTS 2021 and proceed to a [repository re-index]({{page page='elasticsearch-setup#reindex'})

Reviewer: ```suggestion
Once the Elasticsearch cluster is upgraded, start Nuxeo LTS 2021 and proceed to a [repository re-index]({{page page='elasticsearch-setup#reindex'}).
```
